core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -220,4 +262,64 @@ public void run() {
          remove(key, tombstone);
       }
    }
+
+   private final class Scheduler implements BiConsumer<Void, Throwable> {
+      final int targetSize;
+      final long maxDelayMillis;
+
+      int initialSize;
+      int previousPostCleanupSize;
+
+      long currentDelayMillis;
+      volatile boolean running;
+
+      private Scheduler(int targetSize, long maxDelayMillis) {
+         this.targetSize = targetSize;
+         this.maxDelayMillis = maxDelayMillis;
+         currentDelayMillis = maxDelayMillis / 2;
+      }
+
+      void onTaskStarted(int size) {
+         running = true;
+         initialSize = size;
+      }
+
+      void onTaskCompleted(int size) {
+         if (currentDelayMillis == 0) {
+            // no delay between cleanup. If we are below the target size, add 100ms delay before cleanup again
+            if (size < targetSize) {
+               currentDelayMillis = Math.min(100, maxDelayMillis);
+               previousPostCleanupSize = size;
+               scheduleWithCurrentDelay();
+               return;
+            }
+         }
+         double tombstoneCreationRate = (initialSize - previousPostCleanupSize) * 1.0 / currentDelayMillis;
```

Comment:
`currentDelayMillis` can be 0 here, if size >= targetSize.

I'm starting to think I should have let you use the initial algorithm, as this is getting complicated, but I'll have another stab :)

I see you've covered 2 corner cases that I missed in my previous suggestion:
1. It's impossible to compute the rate of change when the delay is 0.
2. Sometimes the creation rate is negative.

Unfortunately, I don't like your solution, because I don't like "magic constants" like the 100ms delay step :)

My proposal would be to:
1. Make the minimum delay 1 instead of 0. It shouldn't have any visible effect on the speed of cleaning up tombstones, but it would get rid of the division by zero risk.
2. When the creation rate is negative, I believe we should be using the maximum delay.

There is one more thing that you're touching a bit with they way you're handling the negative creation rate: you're making the changes incremental, instead of using just the creation rate from the last interval you're also using the creation rate from the previous interval (via the current delay).

I've given it some thought and I think if it's probably worth doing something similar every time, not just when the creation rate is negative, but I'd rather not use a constant step, instead we could do an average between the next delay and the last delay.

So the algorithm ends up something like this:

if (size >= targetSize) {
   currentDelay = 1;
} else {
tombstoneCreationRate = (initialSize - previousPostCleanupSize) * 1.0 / currentDelayMillis
if (tombstoneCreationRate <= 0) {
  nextDelay = maxDelayMillis;
} else {
  nextDelay = (targetSize - size) / tombstoneCreationRate;
}
currentDelayMillis = (currentDelayMillis + nextDelay) / 2;
}

I'm going to try it out myself, hopefully it won't end up adding even more complications...

I did consider doing something with the time between the start and the end of the cleanup task as well, since tombstones are also being created during that time, but at this point I decided that would be too much, even for me :)

Reply:
Ok, my experiment is at https://github.com/infinispan/infinispan/compare/main...danberindei:pr_pruivo_t_13446

Let me know what you think @pruivo ...

Reply:
looks good to me. I've copied your changes.